### PR TITLE
Build and package common libraries (for wasm32-wasi) for external consumption

### PR DESCRIPTION
**Pull request by assambar**
Sample releases here - https://github.com/assambar/webassembly-language-runtimes/releases?q=libs&expanded=true

This is not overly complex, but is a huge change as I had to modify assumptions, approaches and variables in many of the build scripts. See details below.

# There is leftover work which I want to address separately:

- Move bundle-wasmlabs to libs-internal and add corresponding build actions
- Discuss and agree on common folder, asset and target naming for libs (e.g. zlib, uuid, libxml2 follow different conventions)
- Add packaging of libpython as part of the python build
- After merging, release all libraries one by one and update remote dependencies to point to official releases (then delete all releases on assambar/webassembly-language-runtimes)
- When we want to package, the make command should be able to deduce the package version from the target, instead of getting it externally via WASMLABS_PACKAGE_VERSION
- Add autoconf example for `libs/examples/zlib_cmake`

# Details about the change:

 - Added `reusable-build-lib.yaml` and `reusable-release-lib.yaml` which get called for all libraries.
 - Implemented libs packaging as part of the make process.
 - Updated all builder containers:
   - Updated WASI_SDK_VERSION to include the full version - `19.0` and added WASI_SDK_ASSET_NAME which is `wasi-sdk-19.0` and gets appended to release
   - Replaced WASI_SDK_ROOT with WASI_SDK_PATH, as most of github seems to follow this practice.
   - Added curl, libtool, tcl to the wasi-builder base image
   - Now, all libs just use the `wasi-builder:19.0`
 - Added recipes to Makefile.helpers to reduce repetition
   - create_lib_sub_targets to automatically create targets in `Makefile` that forward to targets in `libs/${LIBNAME}/Makefile`
   - added docs and exposed more build state variables for `build_in_container`
   - create_default_lib_targets to automatically create the targets for different lib versions in `libs/${LIBNAME}/Makefile`
 - Projects which build from this repo, must now provide `wl-env-local.sh` for each target, just like the rest provide `wl-env-repo.sh`
 - Added "bundle-wasmlabs" which packages all currently built libraries
   - Hackish solution that uses a "fake" `libs/bundle-wasmlabs/v0.1.0` folder, but works
   - Will be moved to `libs-internal` subsequently.
 - Added a `libs/examples/zlib_cmake` example app which shows how the libraries can be used by external applications. This will grow with time.
 - Add Makefiles for `icu`, `libpng`, `libxml2`, `oniguruma`, `sqlite`
 - Add GH actions for `libpng`, `libxml2`, `oniguruma`, `zlib`, `uuid`. Note that we consume `icu` as header only for now, so did not release anything.
 - Unified dependency mechanism - in `scripts/build-helpers/wlr_dependencies.sh`, where the function `wlr_dependencies_add` implements it all
 - Unified pkg_config prefixes - in `scripts/build-helpers/pkg_config_tools.sh, ensuring we use empty base prefix, and add stuff to libs/wasm32-wasi in .pc files
 - Unified packaging - in `scripts/wlr_oackage.sh`, which uses WASMLABS_PACKAGE_NAME and WASMLABS_PACKAGE_VERSION to package libs files from `include` and `lib`.
 - Removed the default creation of `bin`, `include`, `lib` for all targets - as they are not always needed. Now `wl-build.sh scripts need to make sure they create the folders to which they will publish.
 - Improved info in `wasmlabs-progress.log`
   - we see what is printed during the build of each dependency
   - we see the important variables before each build
 - Changed how and where dependencies get built or populated - by adding a `WASMLABS_BUILD_TYPE` and ensuring that all deps for a main target go to a separate place - e.g. `build-staging/python/v3.1.11/deps/build-staging` and `build-staging/python/v3.1.11/deps/build-output` will contain all files either locally built or downloaded. In this way there is no overlap between different releases of the same version of say zlib, when used by python or php

### CODE REVIEW

**Comment by assambar** [comment]
I will squash the commits on merge !

File: .github/workflows/reusable-release-lib.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh release create --generate-notes ${{ github.ref_name }} || true
+      - name: Append ${{ inputs.name }} release assets
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          gh release upload ${{ github.ref_name }} \
+            build-output/${{ inputs.name }}-${{ steps.determineVersion.outputs.value }}*.gz
```
Comment (assambar) [inline .github/workflows/reusable-release-lib.yaml:45]: Keep only `build-output/*.gz`

File: scripts/build-helpers/pkg_config_tools.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+if [ "${BASH_SOURCE-}" = "$0" ]; then
+    echo "You must source this script: \$ source $0" >&2
+    return
+fi
+
+export PKG_CONFIG_CONFIGURE_PREFIXES="--prefix= --libdir=\${exec_prefix}/lib/wasm32-wasi"
```
Comment (assambar) [inline scripts/build-helpers/pkg_config_tools.sh:8]: Check if this libdir is really needed.

Comment (assambar) [inline scripts/build-helpers/pkg_config_tools.sh:8]: WLR_CONFIGURE_PREFIXES

File: libs/libpng/v1.6.39/wl-build.sh [hunk 3]
```diff
@@ -30,8 +28,15 @@ export LDFLAGS="${LDFLAGS_WASI} ${LDFLAGS_LIBPNG}"
 
 cd "${WASMLABS_SOURCE_PATH}"
 
+source ${WASMLABS_REPO_ROOT}/scripts/build-helpers/pkg_config_tools.sh
+
 if [[ -z "$WASMLABS_SKIP_CONFIGURE" ]]; then
-    export LIBPNG_CONFIGURE="--prefix="${WASMLABS_OUTPUT}" --enable-static --disable-shared"
+    echo -e "\n\n\n"
```
Comment (assambar) [inline libs/libpng/v1.6.39/wl-build.sh:34]: Leftover

File: libs/libpng/v1.6.39/wl-build.sh [hunk 4]
```diff
@@ -12,10 +12,8 @@ export CFLAGS_CONFIG="-O0"
 export CFLAGS_WASI="--sysroot=${WASI_SYSROOT}"
 export LDFLAGS_WASI="--sysroot=${WASI_SYSROOT}"
 
-ZLIB_INCLUDEDIR=$(pkg-config --variable=includedir zlib)
-ZLIB_LIBDIR=$(pkg-config --variable=libdir zlib)
-CFLAGS_LIBPNG="-I"${ZLIB_INCLUDEDIR} 
-LDFLAGS_LIBPNG="-L"${ZLIB_LIBDIR} 
+CFLAGS_LIBPNG="-I"${WLR_DEPS_INCLUDE}
```
Comment (assambar) [inline libs/libpng/v1.6.39/wl-build.sh:15]: Fix quotes.

Comment (assambar) [reply libs/libpng/v1.6.39/wl-build.sh:15]: Just grep for `"$`

File: libs/uuid/libuuid-1.0.3/wl-build.sh [hunk 5]
```diff
@@ -37,9 +39,8 @@ logStatus "Building... "
 make || exit 1
 
 logStatus "Preparing artifacts... "
-make install \
-    prefix=${WASMLABS_OUTPUT} \
-    libdir=${WASMLABS_OUTPUT}/lib \
-    pkgconfigdir=${WASMLABS_OUTPUT}/lib/pkgconfig
+make install ${PKG_CONFIG_INSTALL_PREFIXES} || exit 1
+
+add_pkg_config_Libs ${WASMLABS_OUTPUT}/lib/wasm32-wasi/pkgconfig/uuid.pc ${LDFLAGS_WASI}
```
Comment (assambar) [inline libs/uuid/libuuid-1.0.3/wl-build.sh:44]: wlr_pkgconfig_add_Libs

File: scripts/wl-env.sh [hunk 6]
```diff
@@ -62,6 +62,13 @@ fi
 export WASMLABS_REPO_ROOT="$(git rev-parse --show-toplevel)"
 export WASMLABS_MAKE=${WASMLABS_REPO_ROOT}/wl-make.sh
 
+if [ "${WASMLABS_BUILD_TYPE}" = "dependency" ]
```
Comment (assambar) [inline scripts/wl-env.sh:65]: Just take some time to improve and simplify the logic here. Or log an issue.

File: scripts/wl-env.sh [hunk 7]
```diff
@@ -70,36 +77,59 @@ then
...
 fi
+echo "Using output ${WASMLABS_OUTPUT}."
 
-export WASMLABS_OUTPUT_BASE=${WASMLABS_REPO_ROOT}/build-output
-export WASMLABS_OUTPUT=${WASMLABS_OUTPUT_BASE}/${WASMLABS_ENV_NAME}${WASMLABS_BUILD_FLAVOR:+-$WASMLABS_BUILD_FLAVOR}
+
+if [ "${WASMLABS_BUILD_TYPE}" = "dependency" ]
+then
+    export WASMLABS_OUTPUT=${WASMLABS_DEPS_ROOT}/build-output
+    echo "!!!!!!!!!!! Building ${WASMLABS_ENV_NAME} as a dependency"
```
Comment (assambar) [inline scripts/wl-env.sh:113]: "!" is not necessary

**Review by ereslibre — APPROVED** [review]
LGTM, thanks @assambar

File: scripts/wl-env.sh (hunk 6 above)
Comment (assambar) [reply scripts/wl-env.sh:65]: https://github.com/vmware-labs/webassembly-language-runtimes/issues/75

**Comment by assambar** [comment]
Marked all leftover tasks here for tracking - https://github.com/vmware-labs/webassembly-language-runtimes/issues/76

**Pull request merged**